capa/features/extractors/ts/tools.py

```diff
@@ -0,0 +1,243 @@
+import abc
+import json
+import importlib.resources
+from typing import Dict, List, Tuple, Union, Callable, Iterator, Optional
+from dataclasses import dataclass
+
+from tree_sitter import Node
+
+import capa.features.extractors.ts.signatures
+from capa.features.extractors.script import LANG_CS, LANG_PY
+
+
+@dataclass(frozen=True)
+class BaseNamespace(abc.ABC):
+    """Abstract class for internal representation of the namespace concept, including aliases."""
+
+    name: str
+    node: Node = None
+    alias: str = ""
+
+    def __hash__(self):
+        return hash(self.name)
+
+    def join(self, name: str) -> str:
+        raise NotImplementedError()
+
+
+class CSharpNamespace(BaseNamespace):
+    def join(self, name: str) -> str:
+        """using System; Diagnostics.ProcessStartInfo => System.Diagnostics.ProcessStartInfo"""
+        return LANGUAGE_TOOLKITS[LANG_CS].join_names(self.name, name)
+
+
+class PythonImport(BaseNamespace):
+    def join(self, name: str) -> str:
+        """import subprocess ; subprocess.Popen => subprocess.Popen
+        from threading import Timer (threading.Timer) => Timer
+        """
+        toolkit = LANGUAGE_TOOLKITS[LANG_CS]
+        qualified_names = toolkit.split_name(self.name)
+        if len(qualified_names) < 2:
+            return name
+        return toolkit.join_names(*(qualified_names[:-1] + [name]))
+
+
+class LanguageToolkit:
+    signature_file: str
+    import_signatures: Dict[str, set[str]]
+    method_call_query_type: str
+    property_query_type: str
+    string_delimiters: str
+    integer_prefixes: List[
+        Tuple[Union[str, Tuple[str, ...]], int]
+    ]  # Tends to indicate a number system, e.g. (("0x", "0X"), 16)
+    integer_suffixes: Tuple[str, ...]  # Tends to indicate unsigned (100u) or long (100l) integer literal
+
+    def __init__(self):
+        self.import_signatures = self.load_import_signatures(self.signature_file)
+
+    def load_import_signatures(self, signature_file: str) -> Dict[str, set[str]]:
+        signatures = json.loads(importlib.resources.read_text(capa.features.extractors.ts.signatures, signature_file))
+        return {category: set(namespaces) for category, namespaces in signatures.items()}
+
+    def _is_import(self, name: str) -> bool:
+        return name in self.import_signatures["imports"]
+
+    def is_builtin(self, func: str) -> bool:
+        return func in self.import_signatures["builtins"]
+
+    def join_names(self, *args: str) -> str:
+        return ".".join(args)
+
+    def split_name(self, name: str) -> List[str]:
+        return name.split(".")
+
+    def process_property(self, node: Node, name: str) -> Optional[str]:
+        if self.is_method_call(node):  # yield only p.StartInfo but not p.Start()
+            return None
```

Comment:
can we return `""` to keep the return type consistent?

Reply:
Fixed.